// lightstep-tracer/lib/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
var LOG_INFO = exports.LOG_INFO = 0;
var LOG_WARN = exports.LOG_WARN = 1;
var LOG_ERROR = exports.LOG_ERROR = 2;
var LOG_FATAL = exports.LOG_FATAL = 3;

var LOG_LEVEL_TO_STRING = exports.LOG_LEVEL_TO_STRING = {
    LOG_INFO: 'I',
    LOG_WARN: 'W',
    LOG_ERROR: 'E',
    LOG_FATAL: 'F'
};
var LOG_STRING_TO_LEVEL = exports.LOG_STRING_TO_LEVEL = {
    I: LOG_INFO,
    W: LOG_WARN,
    E: LOG_ERROR,
    F: LOG_FATAL
};

// The report interval for empty reports used to sample the clock skew
var CLOCK_STATE_REFRESH_INTERVAL_MS = exports.CLOCK_STATE_REFRESH_INTERVAL_MS = 350;

var LIGHTSTEP_APP_URL_PREFIX = exports.LIGHTSTEP_APP_URL_PREFIX = 'https://app.lightstep.com';

var JOIN_ID_PREFIX = exports.JOIN_ID_PREFIX = 'join:';

var LS_META_EVENT_KEY = exports.LS_META_EVENT_KEY = 'lightstep.meta_event';
var LS_META_PROPAGATION_KEY = exports.LS_META_PROPAGATION_KEY = 'lightstep.propagation_format';
var LS_META_TRACE_KEY = exports.LS_META_TRACE_KEY = 'lightstep.trace_id';
var LS_META_SPAN_KEY = exports.LS_META_SPAN_KEY = 'lightstep.span_id';
var LS_META_TRACER_GUID_KEY = exports.LS_META_TRACER_GUID_KEY = 'lightstep.tracer_guid';
var LS_META_EXTRACT = exports.LS_META_EXTRACT = 'lightstep.extract_span';
var LS_META_INJECT = exports.LS_META_INJECT = 'lightstep.inject_span';
var LS_META_SP_START = exports.LS_META_SP_START = 'lightstep.span_start';
var LS_META_SP_FINISH = exports.LS_META_SP_FINISH = 'lightstep.span_finish';
var LS_META_TRACER_CREATE = exports.LS_META_TRACER_CREATE = 'lightstep.tracer_create';

var FORMAT_B3 = exports.FORMAT_B3 = 'format.b3';
