// @artilleryio/platform-fargate/lib/data-api/aurora-serverless/migrations/1652207384033-migrate-02-add-testrun-fields.js
const { MigrationInterface, QueryRunner, TableColumn } = require("typeorm");

module.exports = class migrate02AddTestrunFields1652207384033 {
    static name = 'migrate02AddTestrunFields1652207384033';

    async up(queryRunner) {
        await queryRunner.addColumn(
            'test_run',
            new TableColumn({
                name: 'aggregateMetricReport',
                type: 'text',
                isNullable: true,
            }));
        await queryRunner.addColumn(
            'test_run',
            new TableColumn({
                name: 'intermediateMetricReports',
                type: 'jsonb',
                isNullable: true,
            }));
        await queryRunner.addColumn(
            'test_run',
            new TableColumn({
                name: 'consoleLog',
                type: 'text',
                isNullable: true,
            }));
    }

    async down(queryRunner) {
        await queryRunner.dropColumn('test_run', 'aggregateMetricReport');
        await queryRunner.dropColumn('test_run', 'intermediateMetricReport');
        await queryRunner.dropColumn('test_run', 'consoleLog');
    }
}
